### PR TITLE
gni/prov: Add gni ATOMIC_AX (AND XOR) support

### LINKED ISSUE #765: Design and implement open_ops interface for AXOR

### PR DESCRIPTION
**Pull request by chuckfossen**
I changed the fi_gni_ops_ep struct to a struct fi_gni_ops_ep
instead of struct fi_gni_ops_domain since fi_gni_ops_ep and
fi_gni_ops_domain are now unique.

Fixes #765

Signed-off-by: Chuck Fossen chuckf@cray.com

@sungeunchoi @ztiffany

### CODE REVIEW

**Comment by sungeunchoi** [comment]
Would like to see what @agontarek thinks...

**Comment by ztiffany** [comment]
I really don't like the idea of putting the GNI post descriptor in the libfabric interface.  It seems like an avoidable issue.  Also, you're only using the amo_cmd field, why not specify each native AMO command with new GNIX_FAB_RQ_\* types?  One for each native AMO.  The only change needed for that would have been to translate the new GNIX_FAB_RQ_ types to amo commands.  No need to pass post descriptors and GNI amo_cmd parameters everywhere.

We have long, short and fetching versions of AAX, right?  Does Andrew need the fetching version?

**Comment by hppritcha** [comment]
This looks good.  However, for a separate performance PR, i'd suggest specializing the `_gnix_atomic` for different types of atomics.  Its now bloated out to 11 arguments.  I'd like to see, for example, a `_gnix_native_atomic`, etc.

**Comment by chuckfossen** [comment]
Do we need the fetching version of AND XOR as well (GNI_FMA_ATOMIC2_FAX)?

**Comment by sungeunchoi** [comment]
I think we should assume that we'll want to eventually support anything we don't have in libfabric.

File: man/fi_gni.7.md [hunk 1]
```diff
@@ -182,13 +189,23 @@ For *FI_GNI_EP_OPS_1*, the currently supported values are:
...
 
+The `native_amo` function allows the user to call GNI native atomics
+that are not implemented in the libfabric API.
+The parameters for native_amo are the same as the fi_atomic function
+but adds the following parameter:
+
+*enum gnix_fab_req_type req_type*
+: The req_type's supported with this call are GNIX_FAB_RQ_NAMO_AX, and
+GNIX_FAB_RQ_NAMO_AX_S.
+
```
Comment (sungeunchoi) [inline man/fi_gni.7.md]: Can you add a short description of the req_type values in parens after each one?  e.g., "(AND and XOR)".

Comment (chuckfossen) [inline man/fi_gni.7.md]: Yes, will do. BTW, I couldn't figure out how to build the man page so I don't know how it looks.

**Comment by agontarek** [comment]
I think this extension looks reasonable. CCE provides intrinsics for both fetching and non-fetching AAX variants. We don't use the fetching variant internally in libpgas, but user codes making use of the intrinsics would fail if a fetching AAX was not available.

File: man/fi_gni.7.md (hunk 1 above)
Comment (sungeunchoi) [inline man/fi_gni.7.md]: You can install pandoc to build the man pages via "make nroff", but as noted in the README, different versions may generate the man pages differently, so you end up with a bit of a mess in your repository.  With small changes like this, I think it's fine to wait till it gets pushed to the upstream and then look at the final version (including the HTML version) after the bot gets to it.  We do a review of the man pages before every release, so it's not a big deal if you forget.

**Comment by chuckfossen** [comment]
This now supports AX 64/32 bit fetching/non-fetching.

**Comment by ztiffany** [comment]
:+1: 

You might want to change your new GNIX_INFOs to GNIX_DEBUG.

**Pull request merged**